plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -95,6 +115,18 @@ public QueryInsightsService(final ClusterSettings clusterSettings, final ThreadP
                 (settings -> validateExporterConfig(type, settings))
             );
         }
+
+        this.searchQueryMetricsEnabled = clusterSettings.get(SEARCH_QUERY_METRICS_ENABLED_SETTING);
+        clusterSettings.addSettingsUpdateConsumer(SEARCH_QUERY_METRICS_ENABLED_SETTING, this::setSearchQueryMetricsEnabled);
+
+        this.metricsRegistry = metricsRegistry;
+    }
+
+    private void setSearchQueryMetricsEnabled(boolean searchQueryMetricsEnabled) {
```

Comment:
As I think about it more, If we want to reuse query insights listener to add records to the blocking queue and then drain the records, we might need to modify the logic to enable/disable features in query insights. Currently we disable all the async processing in query insights service when all top n services are disabled, because they are all the services we have now. 

We might need to create a new function in queryInsightsService, to disable/enable the async processing depending on if top n services and categorization service are enabled. Otherwise if we disable all top n queries services, the categorization service will stop working.

Reply:
Will [adding this in the isEnabled() check](https://github.com/opensearch-project/OpenSearch/pull/14528/files#diff-529ce87837ab403eb338746abbe3b7fc055c823e2b2fd039e56d193d2f01303bR222) suffice or additional changes needed here?

Reply:
No, we should rewrite the logic for `setEnableTopQueries` in the listener if you intend to reuse the listener for categorization. We might need to check each service in query insights to see if they are all disabled, if so we then disable the whole `queryInsightsService`.